smarty/sampler.py

```diff
@@ -332,38 +335,56 @@ def show_type_matches(self, atom_type_matches):
+    #    """
+    #    if re.match('\$\(\*[=~:\-#](\w+)\)', decorators[0]) != None:
+    #        # There is a bond - two atom types
+    #       result = re.match('\[(.+)\]', atom1type[0])
+    #        basetype_index = random.randint(0, natombasetypes-1)
+    #        (basetype, basetype_typename) = self.atom_basetype[basetype_index]
+    #        new_dec = decorators[0].replace("z", basetype)
+    #        proposed_atomtype = '[' + result.groups(1)[0] + new_dec + ']'
+    #        proposed_typename = atom1type[1] + ' ' + basetype_typename +  ' ' + decorators[1]
+    #    else:
+    #        # Only one atom type
+    #        result = re.match('\[(.+)\]', atom1type[0])
+    #        proposed_atomtype = '[' + result.groups(1)[0] + decorators[0] + ']'
+    #        proposed_typename = atom1type[1] + ' ' + decorators[1]
+    #    return proposed_atomtype, proposed_typename
+
+
+    # I didn't understand why there were base types being introduced in this metho
+    # It should just be decorating the atom, nothing else. 
+    def AtomDecorator(self, atom1type, decorator):
+        """
+        Given an atom and a decorator ammend the SMARTS string with that decorator 
+        Returns at "atom" which is a tuple of [SMARTS, typeName] for the decorated atom
+        """
+        if self.HasAlpha(atom1type):
+            # decorators should go before the $ sign on the atom
+            dollar = atom1type[0].find('$')
+            proposed_atomtype = atom1type[0][:dollar] + decorator[0] + atom3[0][dollar:]
+            proposed_typename = atom1type[1] + decorator[1]
         else:
-            # Only one atom type
-            result = re.match('\[(.+)\]', atom1type[0])
-            proposed_atomtype = '[' + result.groups(1)[0] + decorators[0] + ']'
-            proposed_typename = atom1type[1] + ' ' + decorators[1]
+            # No alpha atom so the decorator goes before the ']'
+            proposed_atomtype = atom1type[0][:-1] + decorator[0] + ']'
+            proposed_typename = atom1type[1] + decorator[1]
         return proposed_atomtype, proposed_typename
 
-    def PickAnAtom(self, natomtypes):
+    def PickAnAtom(self, atomList):
         """
         takes a desired current set of atomtypes (current, base, populated base, or some combination of the 
         above or something else) and pick one at random. (Here, we will typically but perhaps not always want 
         to use currently populated atomtypes, i.e. the atomtypes list, not basetypes (generic types) or 
         used_basetypes (base types which type anything)). A little silly, but still helpful. 
         Returns a tuple corresponding to the selected (smarts, name).
         """
-        atomtype_index = random.randint(0, natomtypes-1)
-        (atomtype, typename) = self.atomtypes[atomtype_index]
-        return (atomtype, typename)
-    
+        atomtype_index = random.randint(0, len(atomList)-1)
+        return atomList[atomtype_index]
```

Comment:
PickAnAtom should be able to find an atom from any atomlist, which list we use should come in the implementation part.